File: convert-to-webp.php
<?php
/**
 * WebP Image Converter Script
 * Converts all JPG, JPEG, and PNG images to WebP format
 * 
 * Usage: Run this script from command line: php convert-to-webp.php
 * Or access via browser: http://localhost/techmech/convert-to-webp.php
 */

// Check if GD library supports WebP
if (!function_exists('imagewebp')) {
    die("Error: WebP support is not available in your PHP installation. Please enable GD extension with WebP support.\n");
}

// Configuration
$baseDir = __DIR__ . '/images';
$quality = 85; // WebP quality (0-100)
$converted = 0;
$skipped = 0;
$errors = 0;

/**
 * Recursively find all image files
 */
function findImages($dir, &$files = []) {
    $items = scandir($dir);
    foreach ($items as $item) {
        if ($item === '.' || $item === '..') continue;
        
        $path = $dir . '/' . $item;
        
        if (is_dir($path)) {
            findImages($path, $files);
        } else {
            $ext = strtolower(pathinfo($item, PATHINFO_EXTENSION));
            if (in_array($ext, ['jpg', 'jpeg', 'png'])) {
                $files[] = $path;
            }
        }
    }
    return $files;
}

/**
 * Convert image to WebP
 */
function convertToWebP($sourcePath, $quality = 85) {
    $ext = strtolower(pathinfo($sourcePath, PATHINFO_EXTENSION));
    $webpPath = preg_replace('/\.(jpg|jpeg|png)$/i', '.webp', $sourcePath);
    
    // Skip if WebP already exists
    if (file_exists($webpPath)) {
        return ['status' => 'skipped', 'message' => 'WebP already exists'];
    }
    
    // Create image resource based on file type
    switch ($ext) {
        case 'jpg':
        case 'jpeg':
            $image = imagecreatefromjpeg($sourcePath);
            break;
        case 'png':
            $image = imagecreatefrompng($sourcePath);
            // Preserve transparency for PNG
            imagealphablending($image, false);
            imagesavealpha($image, true);
            break;
        default:
            return ['status' => 'error', 'message' => 'Unsupported format'];
    }
    
    if (!$image) {
        return ['status' => 'error', 'message' => 'Failed to create image resource'];
    }
    
    // Convert to WebP
    $success = imagewebp($image, $webpPath, $quality);
    
    // Free memory
    imagedestroy($image);
    
    if ($success) {
        // Set same permissions as original
        if (file_exists($sourcePath)) {
            chmod($webpPath, fileperms($sourcePath));
        }
        return ['status' => 'success', 'webp' => $webpPath];
    } else {
        return ['status' => 'error', 'message' => 'Failed to save WebP'];
    }
}

// Start conversion
echo "=== WebP Image Converter ===\n";
echo "Starting conversion...\n\n";

$images = findImages($baseDir);

if (empty($images)) {
    echo "No images found to convert.\n";
    exit;
}

echo "Found " . count($images) . " images to process.\n\n";

foreach ($images as $imagePath) {
    $relativePath = str_replace(__DIR__ . '/', '', $imagePath);
    echo "Processing: $relativePath... ";
    
    $result = convertToWebP($imagePath, $quality);
    
    switch ($result['status']) {
        case 'success':
            $converted++;
            $webpRelative = str_replace(__DIR__ . '/', '', $result['webp']);
            echo "✓ Converted to $webpRelative\n";
            break;
        case 'skipped':
            $skipped++;
            echo "- Skipped (WebP exists)\n";
            break;
        case 'error':
            $errors++;
            echo "✗ Error: " . $result['message'] . "\n";
            break;
    }
}

echo "\n=== Conversion Complete ===\n";
echo "Converted: $converted\n";
echo "Skipped: $skipped\n";
echo "Errors: $errors\n";
echo "\nNext step: Update image references in PHP files to use .webp extension.\n";
?>

File: convert-to-webp.py
#!/usr/bin/env python3
"""
Convert all JPG/PNG images to WebP format
Requires: pip install Pillow
"""

import os
import sys
from pathlib import Path

try:
    from PIL import Image
except ImportError:
    print("Error: Pillow library not found.")
    print("Install it with: pip install Pillow")
    sys.exit(1)

def convert_to_webp(input_path, output_path, quality=85):
    """Convert image to WebP format"""
    try:
        with Image.open(input_path) as img:
            # Convert RGBA to RGB if necessary (for JPG compatibility)
            if img.mode in ('RGBA', 'LA', 'P'):
                # Create white background for transparent images
                rgb_img = Image.new('RGB', img.size, (255, 255, 255))
                if img.mode == 'P':
                    img = img.convert('RGBA')
                rgb_img.paste(img, mask=img.split()[-1] if img.mode in ('RGBA', 'LA') else None)
                img = rgb_img
            elif img.mode != 'RGB':
                img = img.convert('RGB')
            
            # Save as WebP
            img.save(output_path, 'WEBP', quality=quality, method=6)
            return True
    except Exception as e:
        print(f"  ✗ Error: {e}")
        return False

def main():
    base_dir = Path('images')
    if not base_dir.exists():
        print(f"Error: {base_dir} directory not found!")
        sys.exit(1)
    
    quality = 85
    converted = 0
    skipped = 0
    errors = 0
    
    print("=== WebP Image Converter ===\n")
    print(f"Quality: {quality}")
    print(f"Target directory: {base_dir}\n")
    
    # Find all images
    image_extensions = ['.jpg', '.jpeg', '.png']
    images = []
    
    for ext in image_extensions:
        images.extend(base_dir.rglob(f'*{ext}'))
        images.extend(base_dir.rglob(f'*{ext.upper()}'))
    
    if not images:
        print("No images found to convert.")
        sys.exit(0)
    
    print(f"Found {len(images)} images to process.\n")
    
    for img_path in images:
        # Create WebP path
        webp_path = img_path.with_suffix('.webp')
        
        # Skip if WebP already exists
        if webp_path.exists():
            print(f"⏭  Skipping: {img_path.relative_to(base_dir)} (WebP exists)")
            skipped += 1
            continue
        
        print(f"🔄 Converting: {img_path.relative_to(base_dir)}...", end=' ')
        
        if convert_to_webp(img_path, webp_path, quality):
            # Get file sizes
            original_size = img_path.stat().st_size
            webp_size = webp_path.stat().st_size
            reduction = ((original_size - webp_size) / original_size) * 100
            
            print(f"✓ ({reduction:.1f}% smaller)")
            converted += 1
        else:
            errors += 1
    
    print(f"\n=== Conversion Complete ===")
    print(f"✅ Converted: {converted}")
    print(f"⏭  Skipped: {skipped}")
    print(f"✗ Errors: {errors}")
    print(f"\nTotal WebP files created: {converted}")

if __name__ == '__main__':
    main()

